crates/bitwarden-crypto/src/enc_string/symmetric.rs

```diff
@@ -200,11 +232,45 @@ impl EncString {
         Ok(EncString::AesCbc256_HmacSha256_B64 { iv, mac, data })
     }
 
+    pub(crate) fn encrypt_xchacha20_poly1305(
+        data_dec: &[u8],
+        key: &XChaCha20Poly1305Key,
+    ) -> Result<EncString> {
+        let mut protected_header = coset::HeaderBuilder::new().build();
+        protected_header.alg = Some(coset::Algorithm::PrivateUse(cose::XCHACHA20_POLY1305));
+
+        let mut nonce = [0u8; 24];
+        let cose_encrypt0 = coset::CoseEncrypt0Builder::new()
+            .protected(protected_header)
+            .try_create_ciphertext(data_dec, &[], |data, aad| {
+                let ciphertext = crate::xchacha20::encrypt_xchacha20_poly1305(
+                    key.enc_key
+                        .as_slice()
+                        .try_into()
+                        .expect("XChaChaPoly1305 key is 32 bytes long"),
+                    data,
+                    aad,
+                );
```

Comment:
Double checking here: `alg` will be part of the AAD, correct? 

And since the `alg` uniquely maps to this decryption method, it's implied accurate. _However_ I think that means that we need to build in breaking algorithm changes for any subsequent algorithms, which isn't the same as the equality I talk about above.

This is to say that it seems that we should decrypt fail if alg doesn't match the expected cose private use alg for this encryption type.

Reply:
Yes, it is in the protected bucket. From the cose spec:

> This header parameter is used to indicate the algorithm used for the security processing. This header parameter MUST be authenticated where the ability to do so exists. This support is provided by AEAD algorithms or construction (e.g., COSE_Sign and COSE_Mac0). This authentication can be done either by placing the header parameter in the protected-header-parameters bucket or as part of the externally supplied data ([Section 4.3](https://www.rfc-editor.org/rfc/rfc9052.html#Extern_AAD)). The value is taken from the "COSE Algorithms" registry (see [[COSE.Algorithms](https://www.rfc-editor.org/rfc/rfc9052.html#COSE.Algorithms)]).